File: backend/app/models/sync_event.py
# Example using SQLAlchemy ORM for sync event logging
from app.extensions import db
import datetime

class SyncEvent(db.Model):
    """
    ORM model for tracking sync events and logs.
    Stores event type, payload, timestamp, status, and originating device/user.
    Used for audit trail, troubleshooting, and replaying sync events if needed.
    """
    __tablename__ = 'sync_events'

    id = db.Column(db.Integer, primary_key=True)
    event_type = db.Column(db.String, nullable=False)  # e.g., 'stock_update', 'order', etc.
    payload = db.Column(db.JSON, nullable=False)       # The actual data being synced
    timestamp = db.Column(db.DateTime, default=datetime.datetime.utcnow, index=True)  # When the event was created
    status = db.Column(db.String, default='pending', index=True)  # 'pending', 'synced', 'failed', etc.
    device_id = db.Column(db.String, nullable=False, index=True)  # Originating device
    user_id = db.Column(db.String, nullable=True, index=True)     # Originating user (if applicable)

    # Optional: Add an index for faster queries by device and status
    __table_args__ = (
        db.Index('ix_sync_events_device_status', 'device_id', 'status'),
    )

    def __repr__(self):
        return f"<SyncEvent(id={self.id}, type={self.event_type}, status={self.status}, device={self.device_id})>" 

File: frontend/lib/models/sync_event.dart
import 'dart:convert';

/// Model representing a sync event for data synchronization between devices
/// 
/// This model matches the backend SyncEvent structure and includes:
/// - Event type and payload for the actual data being synced
/// - Timestamp for conflict resolution and audit trail
/// - Device and user identification for tracking
/// - Status tracking for sync progress
class SyncEvent {
  final String id;
  final String eventType;
  final Map<String, dynamic> payload;
  final DateTime timestamp;
  final String deviceId;
  final String? userId;
  final String status;
  final bool isCritical;

  /// Create a new sync event
  /// 
  /// [eventType] - Type of event (e.g., 'product_update', 'inventory_update', 'order_create')
  /// [payload] - The actual data being synced
  /// [deviceId] - ID of the originating device
  /// [userId] - ID of the user who triggered the event (optional)
  /// [isCritical] - Whether this event requires immediate sync
  SyncEvent({
    required this.eventType,
    required this.payload,
    required this.deviceId,
    this.userId,
    this.isCritical = false,
    String? id,
    DateTime? timestamp,
    String? status,
  })  : id = id ?? _generateId(),
        timestamp = timestamp ?? DateTime.now(),
        status = status ?? 'pending';

  /// Generate a unique ID for the event
  static String _generateId() {
    return DateTime.now().millisecondsSinceEpoch.toString() + 
           (1000 + (DateTime.now().microsecond % 1000)).toString();
  }

  /// Create a SyncEvent from JSON data (from backend)
  factory SyncEvent.fromJson(Map<String, dynamic> json) {
    return SyncEvent(
      id: json['id']?.toString() ?? '',
      eventType: json['event_type'] ?? '',
      payload: json['payload'] is Map<String, dynamic> 
          ? json['payload'] 
          : jsonDecode(json['payload'].toString()),
      timestamp: json['timestamp'] != null 
          ? DateTime.parse(json['timestamp'])
          : DateTime.now(),
      deviceId: json['device_id'] ?? '',
      userId: json['user_id'],
      status: json['status'] ?? 'pending',
      isCritical: json['is_critical'] ?? false,
    );
  }

  /// Convert SyncEvent to JSON for backend communication
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'event_type': eventType,
      'payload': payload,
      'timestamp': timestamp.toIso8601String(),
      'device_id': deviceId,
      'user_id': userId,
      'status': status,
      'is_critical': isCritical,
    };
  }

  /// Create a copy of this event with updated fields
  SyncEvent copyWith({
    String? id,
    String? eventType,
    Map<String, dynamic>? payload,
    DateTime? timestamp,
    String? deviceId,
    String? userId,
    String? status,
    bool? isCritical,
  }) {
    return SyncEvent(
      id: id ?? this.id,
      eventType: eventType ?? this.eventType,
      payload: payload ?? this.payload,
      timestamp: timestamp ?? this.timestamp,
      deviceId: deviceId ?? this.deviceId,
      userId: userId ?? this.userId,
      status: status ?? this.status,
      isCritical: isCritical ?? this.isCritical,
    );
  }

  /// Mark this event as synced
  SyncEvent markAsSynced() {
    return copyWith(status: 'synced');
  }

  /// Mark this event as failed
  SyncEvent markAsFailed() {
    return copyWith(status: 'failed');
  }

  @override
  String toString() {
    return 'SyncEvent(id: $id, type: $eventType, status: $status, device: $deviceId)';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is SyncEvent && other.id == id;
  }

  @override
  int get hashCode => id.hashCode;
} 